File: go/06_image_safety_check.go
package main

/*
06_image_safety_check.go - Check if image content is safe using LlamaGuard Vision

This example demonstrates how to:
1. Load and encode an image in base64
2. Send it to LlamaGuard for safety analysis
3. Parse the moderation response

LlamaGuard checks images for:
- Violence and graphic content
- Sexual content
- Hate symbols or offensive imagery
- Self-harm content
- Dangerous or illegal activities

Usage:
    go run 06_image_safety_check.go

Environment:
    GROQ_API_KEY - Your Groq API key
*/

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
)

// ContentItem represents text or image content in a message
type ContentItem struct {
	Type     string    `json:"type"`
	Text     string    `json:"text,omitempty"`
	ImageURL *ImageURL `json:"image_url,omitempty"`
}

// ImageURL contains the image data URL
type ImageURL struct {
	URL string `json:"url"`
}

// Message represents a chat message with multi-modal content
type Message struct {
	Role    string        `json:"role"`
	Content []ContentItem `json:"content"`
}

// ChatRequest represents the API request structure
type ChatRequest struct {
	Model     string    `json:"model"`
	Messages  []Message `json:"messages"`
	MaxTokens int       `json:"max_tokens"`
}

// Choice represents a response choice
type Choice struct {
	Message struct {
		Role    string `json:"role"`
		Content string `json:"content"`
	} `json:"message"`
	Index        int    `json:"index"`
	FinishReason string `json:"finish_reason"`
}

// Usage represents token usage statistics
type Usage struct {
	PromptTokens     int `json:"prompt_tokens"`
	CompletionTokens int `json:"completion_tokens"`
	TotalTokens      int `json:"total_tokens"`
}

// ChatResponse represents the API response
type ChatResponse struct {
	ID      string   `json:"id"`
	Object  string   `json:"object"`
	Created int64    `json:"created"`
	Model   string   `json:"model"`
	Choices []Choice `json:"choices"`
	Usage   Usage    `json:"usage"`
	Error   *struct {
		Message string `json:"message"`
		Type    string `json:"type"`
	} `json:"error,omitempty"`
}

// checkImageSafety checks if an image is safe using LlamaGuard Vision
func checkImageSafety(imagePath string) (*ChatResponse, error) {
	// Step 1: Check API key
	apiKey := os.Getenv("GROQ_API_KEY")
	if apiKey == "" {
		return nil, fmt.Errorf("GROQ_API_KEY environment variable not set")
	}

	// Step 2: Load and encode the image
	fmt.Printf("Loading image: %s\n", imagePath)

	// Read the image file
	imageData, err := os.ReadFile(imagePath)
	if err != nil {
		return nil, fmt.Errorf("failed to read image: %w", err)
	}

	// Encode to base64
	imageBase64 := base64.StdEncoding.EncodeToString(imageData)

	// Display image info
	fmt.Printf("Image size: %d bytes\n", len(imageData))
	fmt.Printf("Base64 size: %d characters\n", len(imageBase64))

	// Step 3: Prepare the API request
	request := ChatRequest{
		Model: "meta-llama/llama-guard-4-12b",
		Messages: []Message{
			{
				Role: "user",
				Content: []ContentItem{
					{
						Type: "text",
						Text: "Check this image for safety",
					},
					{
						Type: "image_url",
						ImageURL: &ImageURL{
							URL: fmt.Sprintf("data:image/jpeg;base64,%s", imageBase64),
						},
					},
				},
			},
		},
		MaxTokens: 100, // Safety responses are typically short
	}

	// Convert request to JSON
	jsonData, err := json.Marshal(request)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	// Step 4: Make the API request
	fmt.Println("\nChecking image safety with LlamaGuard...")

	req, err := http.NewRequest("POST", "https://api.groq.com/openai/v1/chat/completions", bytes.NewBuffer(jsonData))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	// Set headers
	req.Header.Set("Authorization", "Bearer "+apiKey)
	req.Header.Set("Content-Type", "application/json")

	// Make the request
	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	// Read the response
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	// Step 5: Parse the response
	var result ChatResponse
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	// Check for API errors
	if result.Error != nil {
		return nil, fmt.Errorf("API error: %s", result.Error.Message)
	}

	return &result, nil
}

// SafetyResult represents the interpretation of a safety check
type SafetyResult struct {
	IsSafe      bool
	Category    string
	Explanation string
}

// interpretSafetyResult interprets the LlamaGuard safety check result
func interpretSafetyResult(response *ChatResponse) SafetyResult {
	// Check if we have a valid response
	if response == nil || len(response.Choices) == 0 {
		return SafetyResult{
			IsSafe:      false,
			Category:    "error",
			Explanation: "No response from safety check",
		}
	}

	// Get the AI's response
	content := strings.ToLower(response.Choices[0].Message.Content)

	// LlamaGuard typically returns "safe" or describes the unsafe category
	isSafe := strings.HasPrefix(content, "safe")

	// Parse the category if unsafe
	if !isSafe {
		// Common unsafe categories from LlamaGuard
		categories := map[string]string{
			"violence":  "Violence or graphic content",
			"sexual":    "Sexual or suggestive content",
			"hate":      "Hate speech or discriminatory content",
			"self-harm": "Self-harm or suicide content",
			"illegal":   "Illegal or dangerous activities",
			"deception": "Deceptive or misleading content",
			"privacy":   "Privacy violation",
			"children":  "Child safety concern",
		}

		// Find which category was flagged
		for key, description := range categories {
			if strings.Contains(content, key) {
				return SafetyResult{
					IsSafe:      false,
					Category:    key,
					Explanation: description,
				}
			}
		}

		// Generic unsafe
		return SafetyResult{
			IsSafe:      false,
			Category:    "unsafe",
			Explanation: content,
		}
	}

	return SafetyResult{
		IsSafe:      true,
		Category:    "safe",
		Explanation: "Image passed safety checks",
	}
}

// fileExists checks if a file exists
func fileExists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}

func main() {
	// Example image paths to check
	testImages := []string{
		"test_image.jpg",    // Default test image
		"../test_image.jpg", // Try parent directory
		"image.jpg",         // Alternative name
	}

	// Find first available image
	var imagePath string
	for _, path := range testImages {
		if fileExists(path) {
			imagePath = path
			break
		}
	}

	if imagePath == "" {
		fmt.Println("No test image found. Please provide an image file.")
		fmt.Println("Expected one of:", strings.Join(testImages, ", "))
		return
	}

	// Check image safety
	response, err := checkImageSafety(imagePath)
	if err != nil {
		fmt.Printf("Error checking image safety: %v\n", err)
		return
	}

	// Interpret results
	result := interpretSafetyResult(response)

	// Display results
	fmt.Println("\n" + strings.Repeat("=", 50))
	fmt.Println("SAFETY CHECK RESULTS")
	fmt.Println(strings.Repeat("=", 50))
	fmt.Printf("Image: %s\n", imagePath)

	// Display status with emoji
	statusEmoji := "✅"
	if !result.IsSafe {
		statusEmoji = "⚠️"
	}
	fmt.Printf("Status: %s %s\n", statusEmoji, strings.ToUpper(result.Category))
	fmt.Printf("Category: %s\n", result.Category)
	fmt.Printf("Details: %s\n", result.Explanation)

	// Show raw response
	if len(response.Choices) > 0 {
		fmt.Printf("\nRaw response: %s\n", response.Choices[0].Message.Content)
	}

	// Display usage statistics
	if response.Usage.TotalTokens > 0 {
		fmt.Println("\n" + strings.Repeat("=", 50))
		fmt.Println("USAGE STATISTICS")
		fmt.Println(strings.Repeat("=", 50))
		fmt.Printf("Prompt tokens: %d\n", response.Usage.PromptTokens)
		fmt.Printf("Completion tokens: %d\n", response.Usage.CompletionTokens)
		fmt.Printf("Total tokens: %d\n", response.Usage.TotalTokens)

		// Calculate approximate cost (LlamaGuard pricing)
		// $0.20 per 1M tokens (both input and output)
		cost := float64(response.Usage.TotalTokens) / 1_000_000 * 0.20
		fmt.Printf("Estimated cost: $%.6f\n", cost)
	}

	// Educational notes
	fmt.Println("\n" + strings.Repeat("=", 50))
	fmt.Println("💡 EDUCATIONAL NOTES")
	fmt.Println(strings.Repeat("=", 50))
	fmt.Println("- LlamaGuard can analyze images for multiple safety categories")
	fmt.Println("- Always check images before processing in production apps")
	fmt.Println("- Consider caching results for frequently checked images")
	fmt.Println("- Combine with text safety checks for complete moderation")
	fmt.Println("- Base64 encoding increases size by ~33%")
	fmt.Println("- Go's encoding/base64 package handles the conversion")
}

File: cpp/06_image_safety_check.cpp
/**
 * ============================================================================
 * Example 6: Image Safety Check - Vision Content Moderation
 * ============================================================================
 *
 * WHAT THIS DEMONSTRATES:
 *   - Using LlamaGuard with vision capabilities
 *   - Checking images for unsafe content
 *   - Combining base64 encoding with safety models
 *
 * WHY IMAGE SAFETY MATTERS:
 *   - User-uploaded images may contain inappropriate content
 *   - AI-generated images should be verified before display
 *   - Legal requirements for content moderation
 *
 * COMPILE:
 *   g++ -std=c++17 -o 06_image_safety_check 06_image_safety_check.cpp -lcurl -I lib/
 *
 * RUN:
 *   ./06_image_safety_check
 *
 * ============================================================================
 */

#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <cstdlib>
#include <curl/curl.h>
#include "json.hpp"

using json = nlohmann::json;

static size_t write_callback(void *contents, size_t size, size_t nmemb, void *userp) {
    std::string *response = static_cast<std::string*>(userp);
    response->append(static_cast<char*>(contents), size * nmemb);
    return size * nmemb;
}

/* Base64 encoding function */
std::string base64_encode(const std::vector<unsigned char> &data) {
    static const char table[] =
        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

    std::string result;
    result.reserve(4 * ((data.size() + 2) / 3));

    size_t i = 0;
    while (i < data.size()) {
        uint32_t octet_a = i < data.size() ? data[i++] : 0;
        uint32_t octet_b = i < data.size() ? data[i++] : 0;
        uint32_t octet_c = i < data.size() ? data[i++] : 0;
        uint32_t triple = (octet_a << 16) + (octet_b << 8) + octet_c;

        result += table[(triple >> 18) & 0x3F];
        result += table[(triple >> 12) & 0x3F];
        result += table[(triple >> 6) & 0x3F];
        result += table[triple & 0x3F];
    }

    size_t mod = data.size() % 3;
    if (mod > 0) {
        result[result.size() - 1] = '=';
        if (mod == 1) result[result.size() - 2] = '=';
    }

    return result;
}

/* Read file into vector */
std::vector<unsigned char> read_file(const std::string &filename) {
    std::ifstream file(filename, std::ios::binary);
    if (!file) {
        throw std::runtime_error("Cannot open file: " + filename);
    }
    return std::vector<unsigned char>(
        std::istreambuf_iterator<char>(file),
        std::istreambuf_iterator<char>()
    );
}

int main() {
    /* Step 1: Get API key */
    const char *api_key_cstr = std::getenv("DEMETERICS_API_KEY");
    if (api_key_cstr == nullptr) {
        std::cerr << "Error: DEMETERICS_API_KEY not set\n";
        return 1;
    }
    std::string api_key(api_key_cstr);

    /* Step 2: Read and encode image */
    std::string image_path = "../test_image.jpg";
    std::cout << "=== IMAGE SAFETY CHECKER ===\n";
    std::cout << "Using LlamaGuard Vision to check image safety\n\n";
    std::cout << "Checking image: " << image_path << "\n";

    std::vector<unsigned char> image_data;
    try {
        image_data = read_file(image_path);
    } catch (const std::exception &e) {
        std::cerr << "Error: " << e.what() << "\n";
        std::cerr << "Make sure test_image.jpg exists in parent directory\n";
        return 1;
    }

    std::string image_base64 = base64_encode(image_data);
    std::cout << "Image size: " << image_data.size() << " bytes\n";
    std::cout << "Encoded length: " << image_base64.size() << " characters\n\n";

    /* Step 3: Initialize curl */
    curl_global_init(CURL_GLOBAL_ALL);
    CURL *curl = curl_easy_init();
    if (curl == nullptr) {
        curl_global_cleanup();
        return 1;
    }

    /* Step 4: Build JSON request with image for LlamaGuard */
    std::string data_url = "data:image/jpeg;base64," + image_base64;

    json request_body = {
        {"model", "meta-llama/llama-guard-4-12b"},
        {"messages", {
            {
                {"role", "user"},
                {"content", {
                    {
                        {"type", "image_url"},
                        {"image_url", {{"url", data_url}}}
                    }
                }}
            }
        }},
        {"max_tokens", 100}
    };

    std::string json_payload = request_body.dump();

    /* Step 5: Set up headers */
    struct curl_slist *headers = nullptr;
    headers = curl_slist_append(headers, "Content-Type: application/json");
    std::string auth_header = "Authorization: Bearer " + api_key;
    headers = curl_slist_append(headers, auth_header.c_str());

    /* Step 6: Configure and perform request */
    std::string response_data;
    curl_easy_setopt(curl, CURLOPT_URL, "https://api.demeterics.com/groq/v1/chat/completions");
    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, json_payload.c_str());
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response_data);

    std::cout << "Analyzing image for safety...\n\n";
    CURLcode res = curl_easy_perform(curl);

    if (res != CURLE_OK) {
        std::cerr << "Error: " << curl_easy_strerror(res) << "\n";
    } else {
        /* Step 7: Parse and display result */
        try {
            json response = json::parse(response_data);

            if (response.contains("choices") && !response["choices"].empty()) {
                std::string result = response["choices"][0]["message"]["content"];

                if (result.substr(0, 4) == "safe") {
                    std::cout << "Result: SAFE\n";
                    std::cout << "The image does not contain unsafe content.\n";
                } else {
                    std::cout << "Result: UNSAFE\n";
                    std::cout << "Raw response: " << result << "\n";
                }
            }
        } catch (const json::exception &e) {
            std::cerr << "Error parsing response: " << e.what() << "\n";
        }
    }

    /* Step 8: Clean up */
    curl_slist_free_all(headers);
    curl_easy_cleanup(curl);
    curl_global_cleanup();

    /* Educational info */
    std::cout << "\n=== IMAGE SAFETY CATEGORIES ===\n\n";
    std::cout << "LlamaGuard checks images for:\n";
    std::cout << "- Violence and gore\n";
    std::cout << "- Adult/sexual content\n";
    std::cout << "- Hate symbols\n";
    std::cout << "- Dangerous activities\n";
    std::cout << "- Child safety issues\n\n";

    std::cout << "=== WHEN TO USE IMAGE SAFETY ===\n\n";
    std::cout << "1. USER UPLOADS: Check all user-submitted images\n";
    std::cout << "2. AI GENERATION: Verify AI-generated images before display\n";
    std::cout << "3. CONTENT FEEDS: Filter images in social features\n";
    std::cout << "4. AVATARS: Validate profile pictures\n";

    return 0;
}
